middleware/metrics_test.go

```diff
@@ -0,0 +1,138 @@
+package middleware
+
+import (
+	"io"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/gorilla/mux"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	"github.com/stretchr/testify/assert"
+)
+
+var tR = prometheus.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "http_requests_total",
+		Help: "Number of get requests.",
+	},
+	[]string{"path"},
+)
+
+var rS = prometheus.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "response_status",
+		Help: "Status of HTTP response",
+	},
+	[]string{"status"},
+)
+
+var hD = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Name: "http_response_time_seconds",
+	Help: "Duration of HTTP requests.",
+}, []string{"path"})
+
+func TestPrometheusRegisters(t *testing.T) {
+	assert := assert.New(t)
+
+	t.Log("Registering Prometheus Metrics")
+
+	prometheus.Register(tR)
+	prometheus.Register(responseStatus)
+	prometheus.Register(httpDuration)
+
+	tR.WithLabelValues("firstLabel").Inc()
+	tR.WithLabelValues("secondLabel").Inc()
+	tR.WithLabelValues("thirdLabel").Inc()
+	tR.WithLabelValues("thirdLabel").Inc()
+
+	rS.WithLabelValues("fourthlabel").Inc()
+	rS.WithLabelValues("fifthlabel").Inc()
+	rS.WithLabelValues("sixthlabel").Inc()
+
+	hD.WithLabelValues("seventhlabel").Observe(1)
+	hD.WithLabelValues("eighthlabel").Observe(2)
+	hD.WithLabelValues("ninelabel").Observe(2)
+
+	// tR collected three metrics
+	assert.Equal(3, testutil.CollectAndCount(tR))
+	// responseStatus collected three metrics
+	assert.Equal(3, testutil.CollectAndCount(rS))
+	// httpDuration collected three metrics
+	assert.Equal(3, testutil.CollectAndCount(hD))
+}
+
+func TestPrometheusMiddlewareAttached(t *testing.T) {
+	assert := assert.New(t)
+
+	t.Log("Launching server in go routine")
+	go func() {
+		t.Log("Defining router and registering prometheus targets")
+		router := mux.NewRouter()
+		router.Use(PrometheusMiddleware)
+		RegisterPrometheus()
+
+		router.Path("/metrics").Handler(promhttp.Handler())
+
+		t.Log("Launching server")
+		if err := http.ListenAndServe(":3030", router); err != nil {
+			t.Log("Panic! ListenAndServe: " + err.Error())
+			panic("ListenAndServe: " + err.Error())
```

Comment:
nit: Let's not directly panic wherever possible.

Reply:
Is a `log.Fatal` equivalent prefferrable to a panic?

Reply:
Panic is generally just bad in general in go, and it's a bad on me to have included one.  Generally [effective go recommends against panic.](https://golang.org/doc/effective_go#panic)

So I consider this not a nit, and will look in to fixing it. I think a t.FailNow() is much better here.

Reply:
t.failnow() won't work inside a goroutine, so I moved to an assert,fail with a go routine exit via `runtime.Goexit()`